### PR TITLE
Make inferred_outlives_crate return Clause

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 1]
```diff
@@ -1360,18 +1360,24 @@ fn early_bound_lifetimes_from_generics<'a, 'tcx: 'a>(
...
     let mut result = tcx.explicit_predicates_of(def_id);
     debug!("predicates_defined_on: explicit_predicates_of({:?}) = {:?}", def_id, result,);
-    let inferred_outlives = tcx.inferred_outlives_of(def_id);
+    let inferred_outlives: Vec<_> = tcx
+        .inferred_outlives_of(def_id)
+        .iter()
+        .map(|(clause, span)| {
+            (ty::Binder::dummy(ty::PredicateKind::Clause(*clause)).to_predicate(tcx), *span)
+        })
+        .collect();
```
Comment (spastorino) [inline compiler/rustc_hir_analysis/src/collect.rs:1369]: I don't like that we need to do this

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/collect.rs:1369]: Can you use `ExactSizeIterator` maybe? Or `Peekable::peek`?

**Comment by oli-obk** [comment]
@rust-timer build 1cbe4ee0b4070e912c551f82b96cce551e4d6f45

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 2]
```diff
@@ -1360,18 +1360,24 @@ fn early_bound_lifetimes_from_generics<'a, 'tcx: 'a>(
 fn predicates_defined_on(tcx: TyCtxt<'_>, def_id: DefId) -> ty::GenericPredicates<'_> {
     let mut result = tcx.explicit_predicates_of(def_id);
     debug!("predicates_defined_on: explicit_predicates_of({:?}) = {:?}", def_id, result,);
-    let inferred_outlives = tcx.inferred_outlives_of(def_id);
+    let inferred_outlives: Vec<_> = tcx
+        .inferred_outlives_of(def_id)
+        .iter()
+        .map(|(clause, span)| {
+            (ty::Binder::dummy(ty::PredicateKind::Clause(*clause)).to_predicate(tcx), *span)
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/collect.rs:1367]: Should be able to just `(*clause).to_predicate()`

File: compiler/rustc_hir_analysis/src/outlives/mod.rs [hunk 3]
```diff
@@ -103,19 +101,13 @@ fn inferred_outlives_crate(tcx: TyCtxt<'_>, (): ()) -> CratePredicatesMap<'_> {
                 |(ty::OutlivesPredicate(kind1, region2), &span)| {
                     match kind1.unpack() {
                         GenericArgKind::Type(ty1) => Some((
-                            ty::Binder::dummy(ty::PredicateKind::Clause(ty::Clause::TypeOutlives(
-                                ty::OutlivesPredicate(ty1, *region2),
-                            )))
-                            .to_predicate(tcx),
+                            ty::Clause::TypeOutlives(ty::OutlivesPredicate(ty1, *region2))
+                                .to_predicate(tcx),
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/outlives/mod.rs:105]: This `to_predicate` is not necessary anymore

File: compiler/rustc_hir_analysis/src/collect.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_hir_analysis/src/collect.rs]: Fixed by creating the iterator inside the check.

**Comment by spastorino** [comment]
@oli-obk rebased and addressed the comments you've made and also fixed the iterator issue.

**Pull request merged**